src/libfsm/dfavm.c

```diff
@@ -0,0 +1,1692 @@
+//
+//   DD=11 is reserved for future use
+//
+
+#define MASK_OP   0x18
+#define MASK_CMP  0xE0
+#define MASK_END  0x01
+#define MASK_DEST 0x03
+
+enum dfavm_instr_bits {
+	// Stop the VM, mark match or failure
+	VM_OP_STOP   = 0,
+
+	// Start of each state: fetch next character.  Indicates
+	// match / fail if EOS.
+	VM_OP_FETCH  = 1,
+
+	// Branch to another state
+	VM_OP_BRANCH = 2,
+};
+
+enum dfavm_cmp_bits {
+	VM_CMP_ALWAYS = 0,
+	VM_CMP_LT     = 1,
+	VM_CMP_LE     = 2,
+	VM_CMP_GE     = 3,
+	VM_CMP_GT     = 4,
+	VM_CMP_EQ     = 5,
+	VM_CMP_NE     = 6,
+};
+
+enum dfavm_end_bits {
+	VM_END_FAIL = 0,
+	VM_END_SUCC = 1,
+};
+
+enum dfavm_dest_bits {
+	VM_DEST_NONE  = 0,
+	VM_DEST_SHORT = 1,  // 11-bit dest
+	VM_DEST_NEAR  = 2,  // 18-bit dest
+	VM_DEST_FAR   = 3,  // 32-bit dest
+};
+
+struct dfavm_op {
+	struct dfavm_op *next;
+	uint32_t count;
+	uint32_t offset;
+
+	uint32_t num_incoming; // number of branches to this instruction
+
+	enum dfavm_cmp_bits cmp;
+	enum dfavm_instr_bits instr;
+
+	union {
+		struct {
+			unsigned state;
+			enum dfavm_end_bits end_bits;
+		} fetch;
+
+		struct {
+			enum dfavm_end_bits end_bits;
+		} stop;
+
+		struct {
+			struct dfavm_op *dest_arg;
+			enum dfavm_dest_bits dest;
+			uint32_t dest_state;
+			int32_t  rel_dest;
+		} br;
+	} u;
+
+	unsigned char cmp_arg;
+	unsigned char num_encoded_bytes;
+	int in_trace;
+};
+
+struct dfavm_op_pool {
+	struct dfavm_op_pool *next;
+
+	unsigned int top;
+	struct dfavm_op ops[1024];
+};
+
+struct dfavm_assembler {
+	struct dfavm_op_pool *pool;
+	struct dfavm_op *freelist;
+
+	struct dfavm_op **ops;
+	struct dfavm_op *linked;
+
+	size_t nstates;
+	size_t start;
+
+	uint32_t nbytes;
+	uint32_t count;
+};
+
+struct fsm_dfavm {
+	unsigned char *ops;
+	uint32_t len;
+};
+
+static const char *
+cmp_name(int cmp) {
+	switch (cmp) {
+	case VM_CMP_ALWAYS: return "";   break;
+	case VM_CMP_LT:     return "LT"; break;
+	case VM_CMP_LE:     return "LE"; break;
+	case VM_CMP_EQ:     return "EQ"; break;
+	case VM_CMP_GE:     return "GE"; break;
+	case VM_CMP_GT:     return "GT"; break;
+	case VM_CMP_NE:     return "NE"; break;
+	default:            return "??"; break;
+	}
+}
+
+static void
+print_op(FILE *f, struct dfavm_op *op)
+{
+	const char *cmp;
+	int nargs = 0;
+
+	cmp = cmp_name(op->cmp);
+
+	fprintf(f, "[%4lu] %1lu\t", (unsigned long)op->offset, (unsigned long)op->num_encoded_bytes);
+
+	switch (op->instr) {
+	case VM_OP_FETCH:
+		fprintf(f, "FETCH%c%s",
+			(op->u.fetch.end_bits == VM_END_FAIL) ? 'F' : 'S',
+			cmp);
+		break;
+
+	case VM_OP_STOP:
+		fprintf(f, "STOP%c%s",
+			(op->u.stop.end_bits == VM_END_FAIL) ? 'F' : 'S',
+			cmp);
+		break;
+
+	case VM_OP_BRANCH:
+		{
+			char dst;
+			switch (op->u.br.dest) {
+			case VM_DEST_NONE:  dst = 'Z'; break;
+			case VM_DEST_SHORT: dst = 'S'; break;
+			case VM_DEST_NEAR:  dst = 'N'; break;
+			case VM_DEST_FAR:   dst = 'F'; break;
+			default:            dst = '!'; break;
+			}
```

Comment:
I do agree that the near/far jumps are probably overkill, but not strongly :)

Reply:
As soon as I have a performance test suite, I'm going to try both 16-bit and 32-bit fixed width bytecodes for the VM.